File: ios/Sendro/Core/Notifier.swift
//
//  Notifier.swift
//  Sendro
//
//  Local notifications (UNUserNotificationCenter) — no push server, no
//  entitlement, no APNs. Everything here is scheduled by the app itself with
//  a nil trigger, i.e. "show this now".
//
//  BEST-EFFORT BY DESIGN: iOS suspends and eventually terminates a
//  backgrounded app. Sendro's outbox long-poll only runs while the app is
//  running (or briefly backgrounded), so these notifications fire for things
//  that arrive in that window — they are NOT push, and the app must never
//  promise "you'll be notified even if Sendro is closed".
//
//  Privacy: a §11 text message NEVER puts its text in a notification body.
//  The body says "Sent you text" and nothing else.
//

import Foundation
import UserNotifications
import Combine

/// Where a notification tap should land.
enum NotificationRoute: String {
    case receive
    case send
    case library
}

/// Which surface the user is looking at right now — used to suppress banners
/// for things that are already visible.
enum AppSurface: String {
    case receive
    case send
    case library
    /// The full-screen Flight view is covering everything else.
    case flight
}

final class Notifier: NSObject, ObservableObject, UNUserNotificationCenterDelegate {

    /// Set when the user taps a notification; RootView consumes and clears it.
    @Published var pendingRoute: NotificationRoute?

    private let settings: Settings

    /// Guards the presentation context, which is written on the main thread
    /// and read on UNUserNotificationCenter's delegate queue.
    private let lock = NSLock()
    private var appActive = true
    private var surface: AppSurface = .receive

    private static let didAskKey = "sendro.notificationsAsked"

    init(settings: Settings) {
        self.settings = settings
        super.init()
        UNUserNotificationCenter.current().delegate = self
    }

    // MARK: Context (called from the UI)

    func updateContext(active: Bool, surface: AppSurface) {
        lock.lock()
        self.appActive = active
        self.surface = surface
        lock.unlock()
    }

    private func snapshot() -> (active: Bool, surface: AppSurface) {
        lock.lock()
        defer { lock.unlock() }
        return (appActive, surface)
    }

    // MARK: Authorization

    /// Asked once, at the first moment it makes sense: right after a pairing
    /// succeeds. Never at launch — a permission sheet before the user has
    /// anything to be notified about is the classic way to get denied.
    func requestAuthorizationAfterPairing() {
        let defaults = UserDefaults.standard
        guard !defaults.bool(forKey: Self.didAskKey) else { return }
        let center = UNUserNotificationCenter.current()
        let key = Self.didAskKey
        center.getNotificationSettings { current in
            guard current.authorizationStatus == .notDetermined else {
                defaults.set(true, forKey: key)
                return
            }
            center.requestAuthorization(options: [.alert, .sound, .badge]) { _, _ in
                defaults.set(true, forKey: key)
            }
        }
    }

    // MARK: Posting

    /// Offers landed in the outbox poll.
    func notifyIncomingOffers(count: Int, senderName: String) {
        guard settings.notifyTransfers, count > 0 else { return }
        guard shouldPost(for: .receive) else { return }
        let body = count == 1
            ? "1 file is waiting for you."
            : "\(count) files are waiting for you."
        post(title: "\(senderName) wants to send \(count == 1 ? "a file" : "\(count) files")",
             body: body,
             route: .receive,
             identifier: "sendro.offers")
    }

    /// A transfer finished (or failed) while the user wasn't watching it.
    func notifyTransferFinished(fileName: String, savedTo: String?) {
        guard settings.notifyTransfers else { return }
        guard shouldPost(for: .library) else { return }
        let destination = savedTo == "photos" ? "Photos" : "Files"
        post(title: "Saved to \(destination)",
             body: fileName,
             route: .library,
             identifier: "sendro.done.\(fileName)")
    }

    func notifyTransferFailed(fileName: String, reason: String) {
        guard settings.notifyTransfers else { return }
        guard shouldPost(for: .receive) else { return }
        post(title: "Transfer failed",
             body: "\(fileName) — \(reason)",
             route: .receive,
             identifier: "sendro.failed.\(fileName)")
    }

    /// §11 text. The text itself is deliberately absent from the payload.
    func notifyMessage(senderName: String) {
        guard settings.notifyMessages else { return }
        guard shouldPost(for: .receive) else { return }
        post(title: senderName,
             body: "Sent you text",
             route: .receive,
             identifier: "sendro.message")
    }

    /// Don't post at all when the user is already looking at the surface the
    /// notification would send them to.
    private func shouldPost(for route: NotificationRoute) -> Bool {
        let state = snapshot()
        guard state.active else { return true }
        return !surfaceShows(route, state.surface)
    }

    private func surfaceShows(_ route: NotificationRoute, _ surface: AppSurface) -> Bool {
        switch route {
        case .receive:  return surface == .receive || surface == .flight
        case .send:     return surface == .send
        case .library:  return surface == .library || surface == .flight
        }
    }

    private func post(title: String, body: String, route: NotificationRoute, identifier: String) {
        let content = UNMutableNotificationContent()
        content.title = title
        content.body = body
        content.sound = .default
        content.userInfo = ["route": route.rawValue]
        let request = UNNotificationRequest(identifier: "\(identifier).\(UUID().uuidString)",
                                            content: content,
                                            trigger: nil)
        UNUserNotificationCenter.current().add(request)
    }

    // MARK: UNUserNotificationCenterDelegate

    func userNotificationCenter(_ center: UNUserNotificationCenter,
                                willPresent notification: UNNotification,
                                withCompletionHandler completionHandler:
                                    @escaping (UNNotificationPresentationOptions) -> Void) {
        let route = Self.route(from: notification.request.content.userInfo)
        let state = snapshot()
        if state.active, let route, surfaceShows(route, state.surface) {
            completionHandler([])           // already on screen — stay quiet
        } else {
            completionHandler([.banner, .sound])
        }
    }

    func userNotificationCenter(_ center: UNUserNotificationCenter,
                                didReceive response: UNNotificationResponse,
                                withCompletionHandler completionHandler: @escaping () -> Void) {
        let route = Self.route(from: response.notification.request.content.userInfo)
        DispatchQueue.main.async { [weak self] in
            self?.pendingRoute = route
            completionHandler()
        }
    }

    private static func route(from userInfo: [AnyHashable: Any]) -> NotificationRoute? {
        guard let raw = userInfo["route"] as? String else { return nil }
        return NotificationRoute(rawValue: raw)
    }
}

File: android/app/src/main/java/com/sendro/android/core/Notifier.kt
package com.sendro.android.core

import android.Manifest
import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Build
import androidx.core.app.NotificationCompat
import androidx.core.app.NotificationManagerCompat
import com.sendro.android.MainActivity
import com.sendro.android.R

/** Where a notification tap should land. */
enum class NotificationRoute { RECEIVE, SEND, LIBRARY }

/** Which surface the user is looking at right now — used to stay quiet. */
enum class AppSurface { RECEIVE, SEND, LIBRARY, FLIGHT, BACKGROUND }

/**
 * Local notifications. No push server, no FCM, no Play Services — everything
 * here is posted by the app itself, for things that arrive while the app (or
 * its foreground service) is running.
 *
 * PRIVACY (§11.3): a text message NEVER puts its text in a notification body.
 * The body says "Sent you text" and nothing else — the same rule as iOS.
 */
class Notifier(
    private val context: Context,
    private val settings: SettingsStore,
) {

    object Channels {
        const val OFFERS = "sendro.offers"
        const val TRANSFERS = "sendro.transfers"
        const val MESSAGES = "sendro.messages"
        const val PROGRESS = "sendro.progress"
    }

    object Ids {
        const val OFFERS = 1001
        const val MESSAGE = 1002
        const val PROGRESS = 1003
        const val UPDATE = 1004
        const val PAIRING = 1005
        /** Finished/failed transfers get an id derived from the transfer. */
        fun forTransfer(transferId: String): Int = 2000 + (transferId.hashCode() and 0x3FFF)
    }

    @Volatile
    private var surface: AppSurface = AppSurface.BACKGROUND

    private val manager = NotificationManagerCompat.from(context)

    init {
        createChannels()
    }

    fun updateSurface(surface: AppSurface) {
        this.surface = surface
    }

    /**
     * API 33+ requires the runtime POST_NOTIFICATIONS grant; below that a
     * notification always posts. Never throw — a missing grant must degrade
     * to "no banner", never to a crash mid-transfer.
     */
    fun canPost(): Boolean {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.TIRAMISU) return true
        return context.checkSelfPermission(Manifest.permission.POST_NOTIFICATIONS) ==
            PackageManager.PERMISSION_GRANTED
    }

    private fun createChannels() {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O) return
        val service = context.getSystemService(NotificationManager::class.java) ?: return
        val channels = listOf(
            NotificationChannel(
                Channels.OFFERS,
                context.getString(R.string.notification_channel_offers),
                NotificationManager.IMPORTANCE_HIGH,
            ).apply { description = "A paired computer wants to send you files." },
            NotificationChannel(
                Channels.TRANSFERS,
                context.getString(R.string.notification_channel_transfers),
                NotificationManager.IMPORTANCE_DEFAULT,
            ).apply { description = "Finished and failed transfers." },
            NotificationChannel(
                Channels.MESSAGES,
                context.getString(R.string.notification_channel_messages),
                NotificationManager.IMPORTANCE_DEFAULT,
            ).apply {
                description = "Someone sent you text. The text itself is never shown here."
            },
            NotificationChannel(
                Channels.PROGRESS,
                context.getString(R.string.notification_channel_progress),
                // Silent: this one is an ongoing progress bar, not an event.
                NotificationManager.IMPORTANCE_LOW,
            ).apply {
                description = "The ongoing notification that keeps a transfer alive."
                setShowBadge(false)
            },
        )
        channels.forEach { runCatching { service.createNotificationChannel(it) } }
    }

    // -----------------------------------------------------------------------
    // Posting
    // -----------------------------------------------------------------------

    fun notifyIncomingOffers(count: Int, senderName: String) {
        if (!settings.current.notifyTransfers || count <= 0) return
        if (!shouldPost(NotificationRoute.RECEIVE)) return
        post(
            id = Ids.OFFERS,
            channel = Channels.OFFERS,
            title = if (count == 1) "$senderName wants to send a file"
            else "$senderName wants to send $count files",
            body = if (count == 1) "1 file is waiting for you."
            else "$count files are waiting for you.",
            route = NotificationRoute.RECEIVE,
        )
    }

    fun notifyTransferFinished(transferId: String, fileName: String, savedTo: String?) {
        if (!settings.current.notifyTransfers) return
        if (!shouldPost(NotificationRoute.LIBRARY)) return
        val destination = when (savedTo) {
            "photos" -> "Gallery"
            "files" -> "Files"
            else -> "this phone"
        }
        post(
            id = Ids.forTransfer(transferId),
            channel = Channels.TRANSFERS,
            title = "Saved to $destination",
            body = fileName,
            route = NotificationRoute.LIBRARY,
        )
    }

    fun notifyTransferFailed(transferId: String, fileName: String, reason: String) {
        if (!settings.current.notifyTransfers) return
        if (!shouldPost(NotificationRoute.RECEIVE)) return
        post(
            id = Ids.forTransfer(transferId),
            channel = Channels.TRANSFERS,
            title = "Transfer failed",
            body = "$fileName — $reason",
            route = NotificationRoute.RECEIVE,
        )
    }

    /** §11 text. The text itself is deliberately absent from the payload. */
    fun notifyMessage(senderName: String) {
        if (!settings.current.notifyMessages) return
        if (!shouldPost(NotificationRoute.RECEIVE)) return
        post(
            id = Ids.MESSAGE,
            channel = Channels.MESSAGES,
            title = senderName,
            body = "Sent you text",
            route = NotificationRoute.RECEIVE,
        )
    }

    /**
     * §15: a device is asking to pair with this one and needs the six digits
     * typed on its side. The digits are shown, deliberately — they are only
     * useful to someone who can already see this screen, which is exactly the
     * physical-presence property §4 relies on.
     */
    fun notifyPairingRequest(peerName: String, code: String) {
        post(
            id = Ids.PAIRING,
            channel = Channels.OFFERS,
            title = "$peerName wants to pair",
            body = "Type $code on that device.",
            route = NotificationRoute.RECEIVE,
        )
    }

    /** §15: a device finished pairing and may now send to this one. */
    fun notifyPaired(peerName: String) {
        cancel(Ids.PAIRING)
        post(
            id = Ids.PAIRING,
            channel = Channels.TRANSFERS,
            title = "$peerName is paired",
            body = "It can now send files to this device.",
            route = NotificationRoute.RECEIVE,
        )
    }

    fun cancel(id: Int) {
        runCatching { manager.cancel(id) }
    }

    /** Don't post when the user is already looking at where it would send them. */
    private fun shouldPost(route: NotificationRoute): Boolean = when (surface) {
        AppSurface.BACKGROUND -> true
        AppSurface.FLIGHT -> route == NotificationRoute.SEND
        AppSurface.RECEIVE -> route != NotificationRoute.RECEIVE
        AppSurface.SEND -> route != NotificationRoute.SEND
        AppSurface.LIBRARY -> route != NotificationRoute.LIBRARY
    }

    @android.annotation.SuppressLint("MissingPermission")
    private fun post(id: Int, channel: String, title: String, body: String, route: NotificationRoute) {
        // canPost() is the runtime guard lint cannot see through.
        if (!canPost()) return
        val notification = NotificationCompat.Builder(context, channel)
            .setSmallIcon(R.drawable.ic_stat_sendro)
            .setContentTitle(title)
            .setContentText(body)
            .setStyle(NotificationCompat.BigTextStyle().bigText(body))
            .setAutoCancel(true)
            .setCategory(NotificationCompat.CATEGORY_STATUS)
            .setPriority(NotificationCompat.PRIORITY_DEFAULT)
            .setContentIntent(contentIntent(route))
            .build()
        runCatching { manager.notify(id, notification) }
    }

    fun contentIntent(route: NotificationRoute): PendingIntent {
        val intent = Intent(context, MainActivity::class.java).apply {
            action = MainActivity.ACTION_ROUTE
            putExtra(MainActivity.EXTRA_ROUTE, route.name)
            flags = Intent.FLAG_ACTIVITY_SINGLE_TOP or Intent.FLAG_ACTIVITY_CLEAR_TOP
        }
        return PendingIntent.getActivity(
            context,
            route.ordinal,
            intent,
            PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE,
        )
    }

    /**
     * The ongoing notification behind [TransferService]. Built here so the
     * channel, icon and tap target stay in one place.
     */
    fun buildProgressNotification(
        title: String,
        text: String,
        progressPercent: Int?,
        indeterminate: Boolean,
    ): Notification {
        val builder = NotificationCompat.Builder(context, Channels.PROGRESS)
            .setSmallIcon(R.drawable.ic_stat_sendro)
            .setContentTitle(title)
            .setContentText(text)
            .setOngoing(true)
            .setOnlyAlertOnce(true)
            .setSilent(true)
            .setCategory(NotificationCompat.CATEGORY_PROGRESS)
            .setPriority(NotificationCompat.PRIORITY_LOW)
            .setForegroundServiceBehavior(NotificationCompat.FOREGROUND_SERVICE_IMMEDIATE)
            .setContentIntent(contentIntent(NotificationRoute.RECEIVE))
        if (indeterminate) {
            builder.setProgress(0, 0, true)
        } else if (progressPercent != null) {
            builder.setProgress(100, progressPercent.coerceIn(0, 100), false)
        }
        return builder.build()
    }
}
